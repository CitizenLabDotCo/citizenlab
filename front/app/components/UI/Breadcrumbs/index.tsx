import React from 'react';
import { colors } from 'utils/styleUtils';
import styled from 'styled-components';

// components
import { Box, Text } from '@citizenlab/cl2-component-library';
import Link from 'utils/cl-router/Link';

const StyledLink = styled(Link)`
  color: ${colors.textSecondary};
  &:hover {
    border-bottom: 2px solid ${colors.textSecondary};
    color: inherit;
    margin-bottom: -2px;
  }
`;

type TBreadcrumb = {
  label: string;
  linkTo?: string;
};

export type TBreadcrumbs = TBreadcrumb[];

interface Props {
  breadcrumbs: TBreadcrumbs;
}

const Breadcrumbs = ({ breadcrumbs }: Props) => {
  if (breadcrumbs.length === 0) {
    return null;
  }

  return (
    <Box display="flex">
      {breadcrumbs.map(({ label, linkTo }, index) => {
        const isLastBreadcrumb = index === breadcrumbs.length - 1;

        return (
          <Box
            key={label}
            display="flex"
            alignItems="center"
            color="textSecondary"
            data-cy={`breadcrumbs-${label}`}
          >
            {linkTo && (
              <Text fontSize="m" as="span">
                <StyledLink to={linkTo}>{label}</StyledLink>
              </Text>
            )}
            {!linkTo && (
              <Text color="textSecondary" fontSize="m" as="span">
                {label}
              </Text>
            )}
            {!isLastBreadcrumb && (
              <Text
                color="borderDark"
                ml="16px"
                as="span"
                mr="16px"
                fontSize="m"
              >
                /
              </Text>
            )}
          </Box>
        );
      })}
    </Box>
  );
};

export default Breadcrumbs;
