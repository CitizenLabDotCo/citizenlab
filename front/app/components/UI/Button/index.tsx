import React, { memo } from 'react';
import { isNilOrError } from 'utils/helperUtils';
import useLocale from 'hooks/useLocale';
import {
  Button,
  ButtonProps,
  ButtonContainerProps as ComponentLibraryButtonContainerProps,
  ButtonStyles as ComponentLibraryButtonStyles,
} from 'cl2-component-library';

interface Props extends Omit<ButtonProps, 'locale'> {}

interface ButtonContainerProps extends ComponentLibraryButtonContainerProps {}

type ButtonStyles = ComponentLibraryButtonStyles;

const ButtonWrapper = memo<Props>((props) => {
  const locale = useLocale();

  if (!isNilOrError(locale)) {
    return <Button {...props} locale={locale} />;
  }

  return null;
});

export default ButtonWrapper;

export { Props, ButtonContainerProps, ButtonStyles };
