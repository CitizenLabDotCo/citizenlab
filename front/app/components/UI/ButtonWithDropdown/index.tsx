import React, { useRef, useState, useEffect, cloneElement } from 'react';
import bowser from 'bowser';
import { trackEventByName } from 'utils/analytics';

// components
import { Dropdown } from 'cl2-component-library';

// styling
import styled from 'styled-components';

const Container = styled.div`
  position: relative;
  display: flex;
  flex-direction: column;
`;

const DropdownWrapper = styled.div`
  width: 100%;
  flex: 0 0 0px;
  position: relative;
  display: flex;
  justify-content: center;
`;

interface Props {
  className?: string;
  buttonComponent: JSX.Element;
  dropdownContent: JSX.Element;
  trackName?: string;
}

const ButtonWithDropdown = ({
  className,
  buttonComponent,
  dropdownContent,
  trackName,
}: Props) => {
  const [buttonWidth, setButtonWidth] = useState(0);
  const ref = useRef<HTMLDivElement | null>(null);
  const [dropdownOpened, setDropdownOpened] = useState(false);

  const onClick = () => {
    toggleDropdown();
    trackName && trackClick(trackName);
  };

  const toggleDropdown = () => {
    setDropdownOpened(!dropdownOpened);
  };

  const trackClick = (trackName: string) => {
    trackEventByName(trackName);
  };

  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(() => {
    if (ref.current) {
      setButtonWidth(ref.current.offsetWidth);
    }
  });

  const button = cloneElement(buttonComponent, {
    onClick,
    ariaExpanded: dropdownOpened,
  });

  return (
    <Container className={className}>
      <div ref={ref}>{button}</div>
      <DropdownWrapper>
        <Dropdown
          top="5px"
          left="0px"
          mobileLeft="0px"
          right={bowser.msie ? '-5px' : 'auto'}
          opened={dropdownOpened}
          onClickOutside={toggleDropdown}
          content={dropdownContent}
          width={`${buttonWidth}px`}
          mobileWidth={`${buttonWidth}px`}
        />
      </DropdownWrapper>
    </Container>
  );
};

export default ButtonWithDropdown;
