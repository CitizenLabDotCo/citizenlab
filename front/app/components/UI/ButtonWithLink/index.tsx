import React, { forwardRef } from 'react';

import {
  Button,
  ButtonProps,
  ButtonContainerProps,
  ButtonStyles,
} from '@citizenlab/cl2-component-library';
import { RouteType } from 'routes';

import Link from 'utils/cl-router/Link';

interface Props extends ButtonProps {
  linkTo?: string | null;
  openLinkInNewTab?: boolean;
  scrollToTop?: boolean;
}

type Ref = HTMLButtonElement;

const ButtonWithLink = forwardRef<Ref, Props>(
  ({ linkTo, openLinkInNewTab, disabled, scrollToTop, ...rest }, ref) => {
    const isExternalLink =
      linkTo &&
      (linkTo.startsWith('http') ||
        linkTo.startsWith('www') ||
        linkTo.startsWith('mailto'));

    const link =
      linkTo && !disabled
        ? isExternalLink
          ? ({
              children,
              ...rest
            }: ButtonProps & React.HTMLAttributes<HTMLAnchorElement>) => (
              <a
                href={linkTo}
                target={openLinkInNewTab ? '_blank' : undefined}
                rel="noreferrer"
                {...rest}
              >
                {children}
              </a>
            )
          : ({
              children,
              ...rest
            }: Omit<ButtonProps, 'as' | 'size'> &
              React.HTMLAttributes<HTMLAnchorElement>) => (
              <Link
                to={linkTo as RouteType}
                target={openLinkInNewTab ? '_blank' : undefined}
                rel="noreferrer"
                scrollToTop={scrollToTop}
                {...rest}
              >
                {children}
              </Link>
            )
        : undefined;

    return <Button as={link} disabled={disabled} {...rest} ref={ref} />;
  }
);

export default ButtonWithLink;

export type { Props, ButtonContainerProps, ButtonStyles, ButtonProps };
