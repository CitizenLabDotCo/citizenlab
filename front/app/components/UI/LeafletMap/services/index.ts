import { init, addTileLayer, changeView } from './setup';

import { addLayersControl, removeLayersControl } from './controls';
import { addLayers, removeLayers, removeLayer } from './layers';
import {
  addMarkersToMap,
  addMarkerClusterGroup,
  getMarkerIcon,
} from './markers';

export default {
  init,
  addTileLayer,
  changeView,
  addLayersControl,
  removeLayersControl,
  addLayers,
  removeLayers,
  removeLayer,
  addMarkersToMap,
  addMarkerClusterGroup,
  getMarkerIcon,
};
