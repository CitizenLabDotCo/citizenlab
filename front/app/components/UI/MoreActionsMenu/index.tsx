// Libraries
import React, { PureComponent } from 'react';
import { removeFocusAfterMouseClick } from 'utils/helperUtils';

// Components
import { Icon, IconNames } from '@citizenlab/cl2-component-library';
import Tippy from '@tippyjs/react';

// Styling
import styled from 'styled-components';
import { colors, fontSizes, media } from 'utils/styleUtils';

const Container = styled.div`
  position: relative;
  display: inline-block;
`;

const MoreOptionsIcon = styled(Icon)<{ color?: string }>`
  fill: ${({ color }) => color || colors.textSecondary};
  transition: all 100ms ease-out;
`;

const MoreOptionsLabel = styled.div`
  color: ${colors.textSecondary};
  font-size: ${fontSizes.s}px;
  line-height: normal;
  font-weight: 400;
  white-space: nowrap;
  margin-left: 10px;
  transition: all 100ms ease-out;

  ${media.tablet`
    display: none;
  `}
`;

const MoreOptionsButton = styled.button`
  min-width: 25px;
  min-height: 20px;
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 0;
  margin: 0;
  border: none;
  background: transparent;
  cursor: pointer;

  &:hover,
  &:focus {
    ${MoreOptionsIcon} {
      fill: #000;
    }

    ${MoreOptionsLabel} {
      color: #000;
    }
  }
`;

const List = styled.div`
  display: flex;
  flex-direction: column;
  align-items: stretch;
  margin-top: 3px;
  margin-bottom: 3px;
`;

const ListItem = styled.button`
  flex: 1 1 auto;
  display: flex;
  align-items: center;
  justify-content: space-between;
  color: ${colors.white};
  font-size: ${fontSizes.s}px;
  font-weight: 400;
  white-space: nowrap;
  padding: 10px;
  border-radius: ${(props) => props.theme.borderRadius};
  cursor: pointer;
  white-space: nowrap;
  transition: all 100ms ease-out;

  & > span {
    margin-right: 5px;
  }

  &:hover,
  &:focus {
    color: white;
    background: ${colors.grey600};
  }
`;

export interface IAction {
  label: string | JSX.Element;
  handler: () => void;
  icon?: IconNames;
  name?: string;
}

export interface Props {
  actions: IAction[];
  // required for a11y
  label: string | JSX.Element;
  showLabel?: boolean;
  className?: string;
  color?: string;
  id?: string;
}

interface State {
  visible: boolean;
}

export default class MoreActionsMenu extends PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      visible: false,
    };
  }

  hideMenu = (event) => {
    event.preventDefault();
    this.setState({ visible: false });
  };

  toggleMenu = (event: React.MouseEvent) => {
    event.preventDefault();
    this.setState(({ visible }) => ({ visible: !visible }));
  };

  handleListItemOnClick =
    (handler: () => void) => (event: React.MouseEvent) => {
      event.preventDefault();
      this.setState({ visible: false });
      handler();
    };

  render() {
    const {
      actions,
      showLabel = true,
      color,
      label,
      className,
      id,
    } = this.props;
    const { visible } = this.state;

    if (!actions || actions.length === 0) {
      return null;
    }

    return (
      <Container className={className || ''}>
        <Tippy
          placement="bottom"
          interactive={true}
          trigger="click"
          duration={[200, 0]}
          content={
            <List className="e2e-more-actions-list">
              {actions.map((action, index) => {
                const { handler, label, icon, name } = action;

                return (
                  <ListItem
                    key={index}
                    onMouseDown={removeFocusAfterMouseClick}
                    onClick={this.handleListItemOnClick(handler)}
                    className={name ? `e2e-action-${name}` : undefined}
                  >
                    {label}
                    {icon && <Icon name={icon} />}
                  </ListItem>
                );
              })}
            </List>
          }
        >
          <MoreOptionsButton
            onMouseDown={removeFocusAfterMouseClick}
            onClick={this.toggleMenu}
            aria-expanded={visible}
            id={id}
            className="e2e-more-actions"
          >
            <MoreOptionsIcon
              title={label}
              name="dots-horizontal"
              color={color}
              ariaHidden={!showLabel}
            />
            {showLabel && <MoreOptionsLabel>{label}</MoreOptionsLabel>}
          </MoreOptionsButton>
        </Tippy>
      </Container>
    );
  }
}
