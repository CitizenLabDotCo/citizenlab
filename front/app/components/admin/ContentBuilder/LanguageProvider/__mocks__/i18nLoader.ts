let translations: Record<string, Record<string, string>> = {
  en: { message: 'Hello' },
  'fr-FR': { message: 'Bonjour' },
};

export const i18nImports = new Proxy(
  {},
  {
    get(_, localePath: string) {
      const match = localePath.match(/\/i18n\/(.+)\.ts/);
      if (!match) {
        return undefined;
      }

      const locale = match[1];
      if (locale in translations) {
        return () =>
          Promise.resolve({
            default: translations[locale],
          });
      }

      return undefined;
    },
  }
);

export const setTranslations = (newTranslations: typeof translations) => {
  translations = newTranslations;
};
