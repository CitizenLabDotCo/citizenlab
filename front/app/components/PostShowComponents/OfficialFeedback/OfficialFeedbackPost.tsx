import React, { PureComponent } from 'react';
import { adopt } from 'react-adopt';
import { isNilOrError } from 'utils/helperUtils';

// typings
import { Locale, Multiloc } from 'typings';

// components
import OfficialFeedbackForm from './OfficialFeedbackForm';
import MoreActionsMenu, { IAction } from 'components/UI/MoreActionsMenu';
import T from 'components/T';
import QuillEditedContent from 'components/UI/QuillEditedContent';

// styles
import { colors, fontSizes, media, isRtl } from 'utils/styleUtils';
import { ScreenReaderOnly } from 'utils/a11y';
import styled from 'styled-components';
import { transparentize } from 'polished';

// i18n
import messages from './messages';
import { FormattedMessage, injectIntl } from 'utils/cl-intl';
import { FormattedDate, InjectedIntlProps } from 'react-intl';
import { getLocalized } from 'utils/i18n';

// services
import {
  IOfficialFeedbackData,
  deleteOfficialFeedbackFromIdea,
  deleteOfficialFeedbackFromInitiative,
} from 'services/officialFeedback';

// resources
import GetLocale, { GetLocaleChildProps } from 'resources/GetLocale';
import GetAppConfigurationLocales, {
  GetAppConfigurationLocalesChildProps,
} from 'resources/GetAppConfigurationLocales';

const Container = styled.div`
  display: flex;
  flex-direction: column;
  border-radius: ${(props: any) => props.theme.borderRadius};
  color: ${({ theme }) => theme.colorText};
  font-size: ${fontSizes.base}px;
  font-weight: 300;
  padding: 30px;
  padding-top: 35px;
  margin-bottom: 15px;

  ${media.smallerThanMinTablet`
    padding: 20px;
    padding-top: 25px;
  `}
`;

const PostContainer = styled(Container)`
  white-space: pre-line;
  background: ${transparentize(0.94, colors.clRedError)};
  position: relative;
`;

const EditFormContainer = styled(Container)`
  background: ${colors.background};
`;

const Body = styled.div`
  margin-bottom: 30px;

  a {
    color: ${colors.clBlueDarker};

    &:hover {
      color: ${colors.clBlueDarker};
    }
  }
`;

const Footer = styled.div`
  display: flex;
  flex-direction: column;
`;

const Author = styled.span`
  color: ${colors.text};
  font-size: ${fontSizes.base}px;
  font-weight: 600;

  ${isRtl`
    text-align: left;
  `}
`;

const DatesPostedEdited = styled.span`
  color: ${colors.text};
  font-size: ${fontSizes.base}px;
  font-weight: 400;
  display: flex;
  align-items: center;

  ${media.smallerThanMinTablet`
    flex-direction: column;
    align-items: flex-start;
  `}

  ${isRtl`
    justify-content: flex-end;
    ${media.smallerThanMinTablet`
        align-items: flex-end;
    `}
 `}
`;

const DatePosted = styled.span``;

const DatesSpacer = styled.span`
  margin-left: 4px;
  margin-right: 4px;

  ${media.smallerThanMinTablet`
    display: none;
  `}
`;

const DateEdited = styled.span`
  ${media.smallerThanMinTablet`
    font-style: italic;
  `}
`;

const StyledMoreActionsMenu = styled(MoreActionsMenu)`
  position: absolute;
  top: 12px;
  right: 15px;

  ${media.smallerThanMinTablet`
    top: 5px;
    right: 5px;
  `}

  ${isRtl`
    right: auto;
    left: 15px;

    ${media.smallerThanMinTablet`
        left: 5px;
    `}
`}
`;

interface DataProps {
  locale: GetLocaleChildProps;
  tenantLocales: GetAppConfigurationLocalesChildProps;
}

interface InputProps {
  editingAllowed: boolean | undefined;
  officialFeedbackPost: IOfficialFeedbackData;
  postType: 'idea' | 'initiative';
  className?: string;
  a11y_pronounceLatestOfficialFeedbackPost?: boolean;
}

interface Props extends InputProps, DataProps {}

interface State {
  showEditForm: boolean;
}

export class OfficialFeedbackPost extends PureComponent<
  Props & InjectedIntlProps,
  State
> {
  constructor(props) {
    super(props);
    this.state = {
      showEditForm: false,
    };
  }

  showEditForm = () => {
    this.setState({ showEditForm: true });
  };

  closeEditForm = () => {
    this.setState({ showEditForm: false });
  };

  deletePost = (postId: string) => () => {
    const { postType } = this.props;

    if (
      window.confirm(
        this.props.intl.formatMessage(messages.deletionConfirmation)
      )
    ) {
      switch (postType) {
        case 'idea':
          deleteOfficialFeedbackFromIdea(postId);
        // eslint-disable-next-line no-fallthrough
        case 'initiative':
          deleteOfficialFeedbackFromInitiative(postId);
      }
    }
  };

  getActions = (postId: string) =>
    [
      {
        label: <FormattedMessage {...messages.editOfficialFeedbackPost} />,
        handler: this.showEditForm,
        name: 'edit',
      },
      {
        label: <FormattedMessage {...messages.deleteOfficialFeedbackPost} />,
        handler: this.deletePost(postId),
        name: 'delete',
      },
    ] as IAction[];

  getPostBodyText = (
    postBodyMultiloc: Multiloc,
    locale: Locale,
    tenantLocales: Locale[]
  ) => {
    const postBodyText = getLocalized(postBodyMultiloc, locale, tenantLocales);
    const processedPostBodyText = postBodyText.replace(
      /<span\sclass="cl-mention-user"[\S\s]*?data-user-id="([\S\s]*?)"[\S\s]*?data-user-slug="([\S\s]*?)"[\S\s]*?>([\S\s]*?)<\/span>/gi,
      '<a class="mention" data-link="/profile/$2" href="/profile/$2">$3</a>'
    );
    return processedPostBodyText;
  };

  render() {
    const {
      editingAllowed,
      officialFeedbackPost,
      locale,
      tenantLocales,
      className,
      a11y_pronounceLatestOfficialFeedbackPost,
    } = this.props;
    const { showEditForm } = this.state;
    const {
      body_multiloc,
      author_multiloc,
      created_at,
      updated_at,
    } = officialFeedbackPost.attributes;

    if (showEditForm && !isNilOrError(locale) && !isNilOrError(tenantLocales)) {
      return (
        <EditFormContainer key={officialFeedbackPost.id}>
          <OfficialFeedbackForm
            locale={locale}
            tenantLocales={tenantLocales}
            formType="edit"
            feedback={officialFeedbackPost}
            onClose={this.closeEditForm}
          />
        </EditFormContainer>
      );
    }

    if (!isNilOrError(locale) && !isNilOrError(tenantLocales)) {
      const formattedPostedOnDate = (
        <FormattedDate
          value={created_at}
          year="numeric"
          month="long"
          day="numeric"
        />
      );

      const formattedUpdatedAtDate = (
        <FormattedDate
          value={updated_at}
          year="numeric"
          month="long"
          day="numeric"
        />
      );

      return (
        <PostContainer
          key={officialFeedbackPost.id}
          className={`e2e-official-feedback-post ${className || ''}`}
        >
          {editingAllowed && (
            <StyledMoreActionsMenu
              ariaLabel={this.props.intl.formatMessage(
                messages.showMoreActions
              )}
              actions={this.getActions(officialFeedbackPost.id)}
            />
          )}

          <ScreenReaderOnly aria-live="polite">
            {a11y_pronounceLatestOfficialFeedbackPost &&
              this.getPostBodyText(body_multiloc, locale, tenantLocales)}
          </ScreenReaderOnly>

          <QuillEditedContent fontWeight={400}>
            <Body className="e2e-official-feedback-post-body">
              <div
                dangerouslySetInnerHTML={{
                  __html: this.getPostBodyText(
                    body_multiloc,
                    locale,
                    tenantLocales
                  ),
                }}
              />
            </Body>
            <Footer>
              <Author className="e2e-official-feedback-post-author">
                <T value={author_multiloc} />
              </Author>

              <DatesPostedEdited>
                <DatePosted>
                  <FormattedMessage
                    {...messages.postedOn}
                    values={{ date: formattedPostedOnDate }}
                  />
                </DatePosted>
                {updated_at && updated_at !== created_at && (
                  <>
                    <DatesSpacer>-</DatesSpacer>
                    <DateEdited>
                      <FormattedMessage
                        {...messages.lastEdition}
                        values={{ date: formattedUpdatedAtDate }}
                      />
                    </DateEdited>
                  </>
                )}
              </DatesPostedEdited>
            </Footer>
          </QuillEditedContent>
        </PostContainer>
      );
    }

    return null;
  }
}

const Data = adopt<DataProps>({
  locale: <GetLocale />,
  tenantLocales: <GetAppConfigurationLocales />,
});

const OfficialFeedbackPostWithIntl = injectIntl<Props>(OfficialFeedbackPost);

export default (inputProps: InputProps) => (
  <Data>
    {(dataProps) => (
      <OfficialFeedbackPostWithIntl {...inputProps} {...dataProps} />
    )}
  </Data>
);
