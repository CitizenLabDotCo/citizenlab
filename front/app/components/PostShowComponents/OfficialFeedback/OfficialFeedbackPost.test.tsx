import React from 'react';
import { shallow } from 'enzyme';
import { IAction } from 'components/UI/MoreActionsMenu';
import * as officialFeedbackSerivce from 'services/officialFeedback';
import { mockOfficialFeedback } from 'services/__mocks__/officialFeedback';
import { OfficialFeedbackPost } from './OfficialFeedbackPost';

jest.mock('services/officialFeedback', () => ({
  deleteOfficialFeedbackFromIdea: jest.fn(),
  deleteOfficialFeedbackFromInitiative: jest.fn(),
}));
jest.mock('./OfficialFeedbackForm', () => 'OfficialFeedbackForm');
jest.mock('components/UI/MoreActionsMenu', () => 'MoreActionsMenu');
jest.mock('components/UI/QuillEditedContent', () => 'QuillEditedContent');
jest.mock('components/T');
jest.mock('utils/cl-intl');

const Intl = require('utils/cl-intl/__mocks__/');
const { intl } = Intl;
const mockOfficialFeedbackPost = mockOfficialFeedback.data[0];

describe('<OfficialFeedbackPost />', () => {
  it('renders correctly for non-admin', () => {
    const wrapper = shallow(
      <OfficialFeedbackPost
        editingAllowed={false}
        officialFeedbackPost={mockOfficialFeedbackPost}
        locale="en"
        tenantLocales={['en', 'nl-BE']}
        intl={intl}
        postType="initiative"
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('renders correctly for admin', () => {
    const wrapper = shallow(
      <OfficialFeedbackPost
        editingAllowed={true}
        officialFeedbackPost={mockOfficialFeedbackPost}
        locale="en"
        tenantLocales={['en', 'nl-BE']}
        intl={intl}
        postType="initiative"
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('intanciates MoreActionsMenu correctly', () => {
    window.confirm = jest.fn(() => true);

    const wrapper = shallow(
      <OfficialFeedbackPost
        editingAllowed={true}
        officialFeedbackPost={mockOfficialFeedbackPost}
        locale="en"
        tenantLocales={['en', 'nl-BE']}
        intl={intl}
        postType="initiative"
      />
    );
    const actions = wrapper
      .find('OfficialFeedbackPost__StyledMoreActionsMenu')
      .prop('actions') as IAction[];
    actions.forEach((action) => action.handler());
    expect(
      officialFeedbackSerivce.deleteOfficialFeedbackFromIdea
    ).toHaveBeenCalledTimes(0);
    expect(
      officialFeedbackSerivce.deleteOfficialFeedbackFromInitiative
    ).toHaveBeenCalledTimes(1);
    expect(wrapper.find('OfficialFeedbackForm'));
  });
});
