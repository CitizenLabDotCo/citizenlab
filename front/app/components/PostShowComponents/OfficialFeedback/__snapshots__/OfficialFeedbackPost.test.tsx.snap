// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OfficialFeedbackPost /> renders correctly for admin 1`] = `
<OfficialFeedbackPost__PostContainer
  className="e2e-official-feedback-post "
  key="feedbackId1"
>
  <OfficialFeedbackPost__StyledMoreActionsMenu
    actions={
      [
        {
          "handler": [Function],
          "label": <FormattedMessage
            defaultMessage="Edit"
            id="app.components.PostComponents.OfficialFeedback.editOfficialFeedbackPost"
          />,
          "name": "edit",
        },
        {
          "handler": [Function],
          "label": <FormattedMessage
            defaultMessage="Delete"
            id="app.components.PostComponents.OfficialFeedback.deleteOfficialFeedbackPost"
          />,
          "name": "delete",
        },
      ]
    }
    label="Show more actions"
  />
  <a11y__ScreenReaderOnly
    aria-live="polite"
  />
  <QuillEditedContent
    fontWeight={400}
  >
    <OfficialFeedbackPost__Body
      className="e2e-official-feedback-post-body"
    >
      <div
        dangerouslySetInnerHTML={
          {
            "__html": "Update: Do Not Keep Calm, please panic now. End of transmission.",
          }
        }
      />
    </OfficialFeedbackPost__Body>
    <OfficialFeedbackPost__Footer>
      <OfficialFeedbackPost__Author
        className="e2e-official-feedback-post-author"
      >
        <T
          value={
            {
              "en": "Testing Official Department",
            }
          }
        />
      </OfficialFeedbackPost__Author>
      <OfficialFeedbackPost__DatesPostedEdited>
        <OfficialFeedbackPost__DatePosted>
          <FormattedMessage
            defaultMessage="Posted on {date}"
            id="app.components.PostComponents.OfficialFeedback.postedOn"
            values={
              {
                "date": <FormattedDate
                  day="numeric"
                  month="long"
                  value="2012-01-01T04:06:07.000Z"
                  year="numeric"
                />,
              }
            }
          />
        </OfficialFeedbackPost__DatePosted>
        <OfficialFeedbackPost__DatesSpacer>
          -
        </OfficialFeedbackPost__DatesSpacer>
        <OfficialFeedbackPost__DateEdited>
          <FormattedMessage
            defaultMessage="Last edited on {date}"
            id="app.components.PostComponents.OfficialFeedback.lastEdition"
            values={
              {
                "date": <FormattedDate
                  day="numeric"
                  month="long"
                  value="2011-01-01T04:26:07.000Z"
                  year="numeric"
                />,
              }
            }
          />
        </OfficialFeedbackPost__DateEdited>
      </OfficialFeedbackPost__DatesPostedEdited>
    </OfficialFeedbackPost__Footer>
  </QuillEditedContent>
</OfficialFeedbackPost__PostContainer>
`;

exports[`<OfficialFeedbackPost /> renders correctly for non-admin 1`] = `
<OfficialFeedbackPost__PostContainer
  className="e2e-official-feedback-post "
  key="feedbackId1"
>
  <a11y__ScreenReaderOnly
    aria-live="polite"
  />
  <QuillEditedContent
    fontWeight={400}
  >
    <OfficialFeedbackPost__Body
      className="e2e-official-feedback-post-body"
    >
      <div
        dangerouslySetInnerHTML={
          {
            "__html": "Update: Do Not Keep Calm, please panic now. End of transmission.",
          }
        }
      />
    </OfficialFeedbackPost__Body>
    <OfficialFeedbackPost__Footer>
      <OfficialFeedbackPost__Author
        className="e2e-official-feedback-post-author"
      >
        <T
          value={
            {
              "en": "Testing Official Department",
            }
          }
        />
      </OfficialFeedbackPost__Author>
      <OfficialFeedbackPost__DatesPostedEdited>
        <OfficialFeedbackPost__DatePosted>
          <FormattedMessage
            defaultMessage="Posted on {date}"
            id="app.components.PostComponents.OfficialFeedback.postedOn"
            values={
              {
                "date": <FormattedDate
                  day="numeric"
                  month="long"
                  value="2012-01-01T04:06:07.000Z"
                  year="numeric"
                />,
              }
            }
          />
        </OfficialFeedbackPost__DatePosted>
        <OfficialFeedbackPost__DatesSpacer>
          -
        </OfficialFeedbackPost__DatesSpacer>
        <OfficialFeedbackPost__DateEdited>
          <FormattedMessage
            defaultMessage="Last edited on {date}"
            id="app.components.PostComponents.OfficialFeedback.lastEdition"
            values={
              {
                "date": <FormattedDate
                  day="numeric"
                  month="long"
                  value="2011-01-01T04:26:07.000Z"
                  year="numeric"
                />,
              }
            }
          />
        </OfficialFeedbackPost__DateEdited>
      </OfficialFeedbackPost__DatesPostedEdited>
    </OfficialFeedbackPost__Footer>
  </QuillEditedContent>
</OfficialFeedbackPost__PostContainer>
`;
